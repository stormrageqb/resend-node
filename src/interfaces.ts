export interface SendEmailData {
  from: string;
  to: string;
  subject?: string;
  text?: string;
  html?: string;
  react?: string;
}
