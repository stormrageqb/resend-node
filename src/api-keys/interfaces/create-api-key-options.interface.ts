import { PostOptions } from '../../common/interfaces';

export interface CreateApiKeyOptions {
  name: string;
}

export interface CreateApiKeyRequestOptions extends PostOptions {}

export interface CreateApiKeyResponse {
  token: string;
}
