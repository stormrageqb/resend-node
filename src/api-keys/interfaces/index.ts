export * from './create-api-key-options.interface';
