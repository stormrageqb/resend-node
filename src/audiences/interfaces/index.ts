export * from './create-audience-options.interface';
export * from './list-audiences.interface';
export * from './get-audience.interface';
export * from './remove-audience.interface';
