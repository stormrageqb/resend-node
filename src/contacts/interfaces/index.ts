export * from './create-contact-options.interface';
export * from './list-contacts.interface';
export * from './get-contact.interface';
export * from './remove-contact.interface';
