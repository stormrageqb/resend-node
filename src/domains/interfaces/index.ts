export * from './create-domain-options.interface';
