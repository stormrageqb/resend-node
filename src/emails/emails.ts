import { render } from '@react-email/render';
import * as React from 'react';
import { Resend } from '../resend';
import {
  CreateEmailOptions,
  CreateEmailRequestOptions,
  CreateEmailResponse,
} from './interfaces';
import { GetEmailResponse } from './interfaces/get-email-options.interface';

export class Emails {
  constructor(private readonly resend: Resend) {}

  async send(
    payload: CreateEmailOptions,
    options: CreateEmailRequestOptions = {},
  ): Promise<CreateEmailResponse> {
    return this.create(payload, options);
  }

  async create(
    payload: CreateEmailOptions,
    options: CreateEmailRequestOptions = {},
  ): Promise<CreateEmailResponse> {
    if (payload.react) {
      payload.html = render(payload.react as React.ReactElement);
      delete payload.react;
    }

    const data = await this.resend.post<CreateEmailResponse>(
      '/emails',
      payload,
      options,
    );
    return data;
  }

  async get(id: string): Promise<GetEmailResponse> {
    const data = await this.resend.get<GetEmailResponse>(`/emails/${id}`);
    return data;
  }
}
