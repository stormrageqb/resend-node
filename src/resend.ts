import { version } from '../package.json';
import { ApiKeys } from './api-keys/api-keys';
import { Audiences } from './audiences/audiences';
import { Batch } from './batch/batch';
import { GetOptions, PostOptions, PutOptions } from './common/interfaces';
import { Domains } from './domains/domains';
import { Emails } from './emails/emails';
import { isResendErrorResponse } from './guards';
import { ErrorResponse, Fetch, Options } from './interfaces';
import { fetch } from 'undici';

const baseUrl = process.env.RESEND_BASE_URL || 'https://api.resend.com';
const userAgent = process.env.RESEND_USER_AGENT || `resend-node:${version}`;

export class Resend {
  private readonly fetch: Fetch = fetch;
  private readonly headers: Headers;

  readonly apiKeys = new ApiKeys(this);
  readonly domains = new Domains(this);
  readonly emails = new Emails(this);
  readonly batch = new Batch(this);
  readonly audiences = new Audiences(this);

  constructor(readonly key?: string, options: Options = {}) {
    if (!key) {
      this.key = process.env.RESEND_API_KEY;

      if (!this.key) {
        throw new Error(
          'Missing API key. Pass it to the constructor `new Resend("re_123")`',
        );
      }
    }

    if (options.fetch) {
      this.fetch = options.fetch;
    }

    this.headers = new Headers({
      Authorization: `Bearer ${this.key}`,
      'User-Agent': userAgent,
      'Content-Type': 'application/json',
    });

    if (options.headers) {
      for (const [key, value] of Object.entries(options.headers)) {
        this.headers.set(key, value);
      }
    }
  }

  async fetchRequest<T>(
    path: string,
    options = {},
  ): Promise<{ data: T | null; error: ErrorResponse | null }> {
    const response = await this.fetch(`${baseUrl}${path}`, options);

    if (!response.ok) {
      const error = await response.json() as any;
      if (isResendErrorResponse(error)) {
        return { data: null, error };
      }

      return { data: null, error };
    }

    const data = await response.json() as any;
    return { data, error: null };
  }

  async post<T>(path: string, entity?: unknown, options: PostOptions = {}) {
    const requestOptions = {
      method: 'POST',
      headers: this.headers,
      body: JSON.stringify(entity),
      ...options,
    };

    return this.fetchRequest<T>(path, requestOptions);
  }

  async get<T>(path: string, options: GetOptions = {}) {
    const requestOptions = {
      method: 'GET',
      headers: this.headers,
      ...options,
    };

    return this.fetchRequest<T>(path, requestOptions);
  }

  async put<T>(path: string, entity: any, options: PutOptions = {}) {
    const requestOptions = {
      method: 'PUT',
      headers: this.headers,
      body: JSON.stringify(entity),
      ...options,
    };

    return this.fetchRequest<T>(path, requestOptions);
  }

  async delete<T>(path: string, query?: unknown) {
    const requestOptions = {
      method: 'DELETE',
      headers: this.headers,
      body: JSON.stringify(query),
    };

    return this.fetchRequest<T>(path, requestOptions);
  }
}
