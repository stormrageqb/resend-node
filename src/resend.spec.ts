import { Resend } from './resend';
import MockAdapater from 'axios-mock-adapter';
import axios from 'axios';
import { CreateEmailOptions } from './emails/interfaces';

const mock = new MockAdapater(axios);
const resend = new Resend('re_924b3rjh2387fbewf823');

describe('Resend', () => {
  afterEach(() => mock.resetHistory());

  it('throws API key error', () => {
    expect(() => new Resend()).toThrowErrorMatchingSnapshot();
  });

  it('sends email', async () => {
    const payload: CreateEmailOptions = {
      from: 'bu@resend.com',
      to: 'zeno@resend.com',
      subject: 'Hello World',
      html: '<h1>Hello world</h1>',
    };
    mock.onPost('/emails', payload).replyOnce(200, {
      id: '1234',
      from: 'bu@resend.com',
      to: 'zeno@resend.com',
      created_at: '123',
    });

    const data = await resend.emails.send(payload);
    expect(data).toMatchInlineSnapshot(`
      {
        "created_at": "123",
        "from": "bu@resend.com",
        "id": "1234",
        "to": "zeno@resend.com",
      }
    `);
  });

  it('sends email with multiple recipients', async () => {
    const payload: CreateEmailOptions = {
      from: 'admin@resend.com',
      to: ['bu@resend.com', 'zeno@resend.com'],
      subject: 'Hello World',
      text: 'Hello world',
    };
    mock.onPost('/emails', payload).replyOnce(200, {
      id: '1234',
      from: 'admin@resend.com',
      to: ['bu@resend.com', 'zeno@resend.com'],
      created_at: '123',
    });

    const data = await resend.emails.send(payload);
    expect(data).toMatchInlineSnapshot(`
      {
        "created_at": "123",
        "from": "admin@resend.com",
        "id": "1234",
        "to": [
          "bu@resend.com",
          "zeno@resend.com",
        ],
      }
    `);
  });

  it('sends email with multiple bcc recipients', async () => {
    const payload: CreateEmailOptions = {
      from: 'admin@resend.com',
      to: 'bu@resend.com',
      bcc: ['foo@resend.com', 'bar@resend.com'],
      subject: 'Hello World',
      text: 'Hello world',
    };
    mock.onPost('/emails', payload).replyOnce(200, {
      id: '1234',
      from: 'admin@resend.com',
      to: 'bu@resend.com',
      bcc: ['foo@resend.com', 'bar@resend.com'],
      created_at: '123',
    });

    const data = await resend.emails.send(payload);
    expect(data).toMatchInlineSnapshot(`
      {
        "bcc": [
          "foo@resend.com",
          "bar@resend.com",
        ],
        "created_at": "123",
        "from": "admin@resend.com",
        "id": "1234",
        "to": "bu@resend.com",
      }
    `);
  });

  it('sends email with multiple cc recipients', async () => {
    const payload: CreateEmailOptions = {
      from: 'admin@resend.com',
      to: 'bu@resend.com',
      cc: ['foo@resend.com', 'bar@resend.com'],
      subject: 'Hello World',
      text: 'Hello world',
    };
    mock.onPost('/emails', payload).replyOnce(200, {
      id: '1234',
      from: 'admin@resend.com',
      to: 'bu@resend.com',
      cc: ['foo@resend.com', 'bar@resend.com'],
      created_at: '123',
    });

    const data = await resend.emails.send(payload);
    expect(data).toMatchInlineSnapshot(`
      {
        "cc": [
          "foo@resend.com",
          "bar@resend.com",
        ],
        "created_at": "123",
        "from": "admin@resend.com",
        "id": "1234",
        "to": "bu@resend.com",
      }
    `);
  });

  it('sends email with multiple replyTo emails', async () => {
    const apiPayload: CreateEmailOptions = {
      from: 'admin@resend.com',
      to: 'bu@resend.com',
      reply_to: ['foo@resend.com', 'bar@resend.com'],
      subject: 'Hello World',
      text: 'Hello world',
    };

    mock.onPost('/emails', apiPayload).replyOnce(200, {
      id: '1234',
      from: 'admin@resend.com',
      to: 'bu@resend.com',
      reply_to: ['foo@resend.com', 'bar@resend.com'],
      created_at: '123',
    });

    const payload: CreateEmailOptions = {
      from: 'admin@resend.com',
      to: 'bu@resend.com',
      reply_to: ['foo@resend.com', 'bar@resend.com'],
      subject: 'Hello World',
      text: 'Hello world',
    };

    const data = await resend.emails.send(payload);
    expect(data).toMatchInlineSnapshot(`
      {
        "created_at": "123",
        "from": "admin@resend.com",
        "id": "1234",
        "reply_to": [
          "foo@resend.com",
          "bar@resend.com",
        ],
        "to": "bu@resend.com",
      }
    `);
  });

  it('can send an email with headers', async () => {
    const payload: CreateEmailOptions = {
      from: 'admin@resend.com',
      headers: {
        'X-Entity-Ref-ID': '123',
      },
      subject: 'Hello World',
      text: 'Hello world',
      to: 'bu@resend.com',
    };
    mock.onPost('/emails', payload).replyOnce(200, {
      id: '1234',
    });

    const data = await resend.emails.send(payload);
    expect(data).toMatchInlineSnapshot(`
      {
        "id": "1234",
      }
    `);
  });
});
