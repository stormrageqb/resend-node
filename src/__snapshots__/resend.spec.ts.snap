// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resend sends email 1`] = `
Object {
  "created_at": "123",
  "from": "bu@resend.com",
  "id": "1234",
  "to": "zeno@resend.com",
}
`;

exports[`Resend sends email with multiple bcc recipients 1`] = `
Object {
  "bcc": Array [
    "foo@resend.com",
    "bar@resend.com",
  ],
  "created_at": "123",
  "from": "admin@resend.com",
  "id": "1234",
  "to": "bu@resend.com",
}
`;

exports[`Resend sends email with multiple cc recipients 1`] = `
Object {
  "cc": Array [
    "foo@resend.com",
    "bar@resend.com",
  ],
  "created_at": "123",
  "from": "admin@resend.com",
  "id": "1234",
  "to": "bu@resend.com",
}
`;

exports[`Resend sends email with multiple recipients 1`] = `
Object {
  "created_at": "123",
  "from": "admin@resend.com",
  "id": "1234",
  "to": Array [
    "bu@resend.com",
    "zeno@resend.com",
  ],
}
`;

exports[`Resend sends email with multiple replyTo emails 1`] = `
Object {
  "created_at": "123",
  "from": "admin@resend.com",
  "id": "1234",
  "reply_to": Array [
    "foo@resend.com",
    "bar@resend.com",
  ],
  "to": "bu@resend.com",
}
`;

exports[`Resend throws API key error 1`] = `"Missing API key. Pass it to the constructor \`new Resend(\\"re_123\\")\`"`;
